engine/common/mod_bmodel.c

```diff
@@ -1881,313 +2029,352 @@ static void Mod_LoadMarkSurfaces( dbspmodel_t *bmod )
-					R_GetBuiltinTexture( REF_ALPHASKY_TEXTURE ) )
-					SetBits( world.flags, FWORLD_SKYSPHERE );
-				continue;
+			// NOTE: We can't load the _luma texture from the WAD as normal because it
+			// doesn't exist there. The original texture is already loaded, but cannot be modified.
+			// Instead, load the original texture again and convert it to luma.
+
+			wadIndex = Mod_FindTextureInWadList( &bmod->wadlist, texture->name, texpath, sizeof(texpath) );
+
+			if( wadIndex >= 0 )
+			{
+				src = FS_LoadFile( texpath, &srcSize, false );
+				bmod->wadlist.wadusage[wadIndex]++;
 			}
 
-			// texture loading order:
-			// 1. from wad
-			// 2. internal from map
+			// OK, loading it from wad or hi-res version
+			texture->fb_texturenum = ref.dllFuncs.GL_LoadTexture( texName, src, srcSize, TF_MAKELUMA );
 
-			// trying wad texture (force while r_wadtextures is 1)
-			if(( r_wadtextures->value && bmod->wadlist.count > 0 ) || ( mt->offsets[0] <= 0 ))
+			if( src )
 			{
-				Q_snprintf( texname, sizeof( texname ), "%s.mip", mt->name );
+				Mem_Free( src );
+			}
+		}
+	}
+#endif // !XASH_DEDICATED
+}
 
-				// check wads in reverse order
-				for( j = bmod->wadlist.count - 1; j >= 0; j-- )
-				{
-					char texpath[MAX_VA_STRING];
+static void Mod_LoadTexture( dbspmodel_t* bmod, int textureIndex )
+{
+	texture_t* texture = NULL;
+	mip_t* mipTex = NULL;
 
-					Q_snprintf( texpath, sizeof( texpath ), "%s.wad/%s", bmod->wadlist.wadnames[j], texname );
+	if( textureIndex < 0 || textureIndex >= loadmodel->numtextures )
+	{
+		return;
+	}
 
-					if( FS_FileExists( texpath, false ))
-					{
-						tx->gl_texturenum = ref.dllFuncs.GL_LoadTexture( texpath, NULL, 0, txFlags );
-						bmod->wadlist.wadusage[j]++; // this wad are really used
-						break;
-					}
-				}
-			}
+	mipTex = Mod_GetMipTexForTexture( bmod, textureIndex );
 
-			// wad failed, so use internal texture (if present)
-			if( mt->offsets[0] > 0 && !tx->gl_texturenum )
-			{
-				// NOTE: imagelib detect miptex version by size
-				// 770 additional bytes is indicated custom palette
-				int	size = (int)sizeof( mip_t ) + ((mt->width * mt->height * 85)>>6);
+	if( !mipTex )
+	{
+		// No data for this texture.
+		// Create default texture (some mods require this).
+		Mod_CreateDefaultTexture( &loadmodel->textures[textureIndex] );
+		return;
+	}
 
-				if( custom_palette ) size += sizeof( short ) + 768;
-				Q_snprintf( texname, sizeof( texname ), "#%s:%s.mip", loadstat.name, mt->name );
-				tx->gl_texturenum = ref.dllFuncs.GL_LoadTexture( texname, (byte *)mt, size, txFlags );
-			}
+	if( mipTex->name[0] == '\0' )
+	{
+		Q_snprintf( mipTex->name, sizeof(mipTex->name), "miptex_%i", textureIndex );
+	}
 
-			// if texture is completely missed
-			if( !tx->gl_texturenum )
-			{
-				Con_DPrintf( S_ERROR "unable to find %s.mip\n", mt->name );
-				tx->gl_texturenum = R_GetBuiltinTexture( REF_DEFAULT_TEXTURE );
-			}
+	texture = (texture_t*)Mem_Calloc( loadmodel->mempool, sizeof(texture_t) );
```

Comment:
`sizeof( texture_t )`

It might be even better if you will use the variable type, like so: `sizeof( *texture )`.